test/extended/router/http2.go

```diff
@@ -172,25 +255,123 @@ var _ = g.Describe("[sig-network-edge][Conformance][Area:Networking][Feature:Rou
+					return true, nil
 				})).NotTo(o.HaveOccurred())
 
 				if tc.expectedGetError != "" {
 					continue
 				}
 
+				o.Expect(resp).ToNot(o.BeNil(), testConfig)
 				o.Expect(resp.StatusCode).To(o.Equal(tc.statusCode), testConfig)
 				o.Expect(resp.Proto).To(o.Equal(tc.frontendProto), testConfig)
 				body, err := ioutil.ReadAll(resp.Body)
 				o.Expect(err).NotTo(o.HaveOccurred(), testConfig)
+				resp.Body.Close()
 				o.Expect(string(body)).To(o.Equal(tc.backendProto), testConfig)
 				o.Expect(resp.Body.Close()).NotTo(o.HaveOccurred())
 			}
 		})
 	})
 })
+
+func getDefaultIngressClusterDomainName(oc *exutil.CLI, timeout time.Duration) (string, error) {
+	var domain string
+
+	if err := wait.PollImmediate(time.Second, timeout, func() (bool, error) {
+		ingress, err := oc.AdminConfigClient().ConfigV1().Ingresses().Get(context.TODO(), "cluster", metav1.GetOptions{})
+		if err != nil {
+			e2e.Logf("Get ingresses.config/cluster failed: %v, retrying...", err)
+			return false, nil
+		}
+		domain = ingress.Spec.Domain
+		return true, nil
+	}); err != nil {
+		return "", err
+	}
+
+	return domain, nil
+}
+
+func getRouterService(oc *exutil.CLI, timeout time.Duration, name string) (*v1.Service, error) {
+	var svc *v1.Service
+
+	if err := wait.PollImmediate(time.Second, timeout, func() (bool, error) {
+		o, err := oc.AdminKubeClient().CoreV1().Services("openshift-ingress").Get(context.Background(), name, metav1.GetOptions{})
+		if err != nil {
+			e2e.Logf("Get router service %s failed: %v, retrying...", name, err)
+			return false, nil
+		}
+		svc = o
+		return true, nil
+	}); err != nil {
+		return nil, err
+	}
+
+	return svc, nil
+}
+
+func getCanaryImage(oc *exutil.CLI) (string, error) {
+	o, err := oc.AdminConfigClient().ConfigV1().ClusterOperators().Get(context.Background(), "ingress", metav1.GetOptions{})
+	if err != nil {
+		return "", err
+	}
+	for _, v := range o.Status.Versions {
+		if v.Name == "canary-server" {
+			return v.Version, nil
+		}
+	}
+	return "", fmt.Errorf("expected to find canary-server version on clusteroperators/ingress")
+}
+
+func resolveHost(oc *exutil.CLI, interval, timeout time.Duration, host string) ([]string, error) {
+	var result []string
+
+	if err := wait.PollImmediate(interval, timeout, func() (bool, error) {
+		addrs, err := net.LookupHost(host)
+		if err != nil {
+			e2e.Logf("error: %v, retrying in %s...", err, interval.String())
+			return false, nil
+		}
+		result = addrs
+		return true, nil
+	}); err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+func resolveHostAsAddress(oc *exutil.CLI, interval, timeout time.Duration, host, expectedAddr string) ([]string, error) {
+	var result []string
+
+	if err := wait.PollImmediate(interval, timeout, func() (bool, error) {
+		addrs, err := net.LookupHost(host)
+		if err != nil {
+			e2e.Logf("error: %v, retrying in %s...", err, interval.String())
+			return false, nil
+		}
+
+		for i := range addrs {
+			if addrs[i] == expectedAddr {
```

Comment:
Why index `addrs` instead of using direct assignment in the `for` statement?

Reply:
Just cheaper.